runtime/shared_common/ClasspathManagerImpl2.cpp

```diff
@@ -1065,29 +1072,37 @@ SH_ClasspathManagerImpl2::validate(J9VMThread* currentThread, ROMClassWrapper* f
 		IDATA prevMatch = ID_NOT_FOUND;
 
 		foundIdentified = -1;
-
-		/* Multiple ClassLoaders could have the same classpath, so search until the matching helperID is found */
-		do {
-			if ((foundIdentified = localValidate_FindIdentified(currentThread, testCP, foundIdentified+1)) != ID_NOT_FOUND) {
-				prevMatch = foundIdentified;
-			}
-		} while ((foundIdentified != ID_NOT_FOUND) && (foundIdentified != compareToID));
-
-		/* This means we have a positive match - testCP is the same classpath as that of the caller classloader */
-		if (foundIdentified==compareToID) {
-			localFoundAtIndex = testCPIndex;
-		} else {
-			/* At this point, foundIdentified will always be ID_NOT_FOUND. This means that testCP is not exactly the same classpath as the caller
-			 * classloader's, but we have identified it and it could still be a valid match. Test to see if we have tried this match before and failed it. */
-			foundIdentified = prevMatch;
-			if (foundIdentified != ID_NOT_FOUND) {
-				if (hasMatchFailedBefore(currentThread, _identifiedClasspaths, compareToID, foundIdentified, testCPIndex, NULL, 0)) {
-					/* trace event is at level 1 and trace exit message is at level 2 as per CMVC 155318/157683 */
-					Trc_SHR_CMI_validate_ExitFailedBefore_Event(currentThread, foundROMClass, compareTo, confirmedEntries);
-					Trc_SHR_CMI_validate_ExitFailedBefore(currentThread);
-					goto _done;
+		if (0 == _cache->enterLocalMutex(currentThread, _identifiedMutex, "identifiedMutex", "validate")) {
+			releaseIdentifiedMutex = true;
+			/* Multiple ClassLoaders could have the same classpath, so search until the matching helperID is found */
+			do {
+				if ((foundIdentified = localValidate_FindIdentified(currentThread, testCP, foundIdentified+1)) != ID_NOT_FOUND) {
+					prevMatch = foundIdentified;
+				}
+			} while ((foundIdentified != ID_NOT_FOUND) && (foundIdentified != compareToID));
+	
+			/* This means we have a positive match - testCP is the same classpath as that of the caller classloader */
+			if (foundIdentified==compareToID) {
+				localFoundAtIndex = testCPIndex;
+			} else {
+				/* At this point, foundIdentified will always be ID_NOT_FOUND. This means that testCP is not exactly the same classpath as the caller
+				 * classloader's, but we have identified it and it could still be a valid match. Test to see if we have tried this match before and failed it. */
+				foundIdentified = prevMatch;
+				if (foundIdentified != ID_NOT_FOUND) {
+					if (hasMatchFailedBefore(currentThread, _identifiedClasspaths, compareToID, foundIdentified, testCPIndex, NULL, 0)) {
+						/* trace event is at level 1 and trace exit message is at level 2 as per CMVC 155318/157683 */
+						Trc_SHR_CMI_validate_ExitFailedBefore_Event(currentThread, foundROMClass, compareTo, confirmedEntries);
+						Trc_SHR_CMI_validate_ExitFailedBefore(currentThread);
+						if (releaseIdentifiedMutex) {
```

Comment:
I don't see why releaseIdentifiedMutex is needed here.

Reply:
Fixed.